core/src/test/scala/integration/kafka/server/ReplicaFetcherThreadFatalErrorTest.scala

```diff
@@ -0,0 +1,145 @@
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import java.util.concurrent.atomic.AtomicBoolean
+
+import kafka.admin.AdminUtils
+import kafka.cluster.BrokerEndPoint
+import kafka.server.ReplicaFetcherThread.{FetchRequest, PartitionData}
+import kafka.utils.{Exit, TestUtils, ZkUtils}
+import kafka.utils.TestUtils.createBrokerConfigs
+import kafka.zk.ZooKeeperTestHarness
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.internals.FatalExitError
+import org.apache.kafka.common.metrics.Metrics
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.FetchResponse
+import org.apache.kafka.common.utils.Time
+import org.junit.{After, Test}
+
+import scala.collection.Map
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+
+class ReplicaFetcherThreadFatalErrorTest extends ZooKeeperTestHarness {
+
+  private var brokers: Seq[KafkaServer] = null
+  @volatile private var shutdownCompleted = false
+
+  @After
+  override def tearDown() {
+    Exit.resetExitProcedure()
+    brokers.foreach(_.shutdown())
+    super.tearDown()
+  }
+
+  /**
+    * Verifies that a follower shuts down if the offset for an `added partition` is out of range and if a fatal
+    * exception is thrown from `handleOffsetOutOfRange`. It's a bit tricky to ensure that there are no deadlocks
+    * when the shutdown hook is invoked and hence this test.
+    */
+  @Test
+  def testFatalErrorInAddPartitions(): Unit = {
+
+    // Unlike `TestUtils.createTopic`, this doesn't wait for metadata propagation as the broker shuts down before
+    // the metadata is propagated.
+    def createTopic(zkUtils: ZkUtils, topic: String): Unit = {
+      AdminUtils.createTopic(zkUtils, topic, partitions = 1, replicationFactor = 2)
+      TestUtils.waitUntilLeaderIsElectedOrChanged(zkUtils, topic, 0)
+    }
+
+    val props = createBrokerConfigs(2, zkConnect)
+    brokers = props.map(KafkaConfig.fromProps).map(config => createServer(config, { params =>
+      import params._
+      new ReplicaFetcherThread(threadName, fetcherId, sourceBroker, config, replicaManager, metrics, time, quotaManager) {
+        override def handleOffsetOutOfRange(topicPartition: TopicPartition): Long = throw new FatalExitError
+        override def addPartitions(partitionAndOffsets: Map[TopicPartition, Long]): Unit =
+          super.addPartitions(partitionAndOffsets.mapValues(_ => -1))
+      }
+    }))
+    createTopic(zkUtils, "topic")
+    TestUtils.waitUntilTrue(() => shutdownCompleted, "Shutdown of follower did not complete", waitTime = 120000)
```

Comment:
Good catch, I had added that while debugging an issue and it should not be needed. Will remove.